modules/service/src/main/scala/lucuma/odb/service/ProgramUserService.scala

```diff
@@ -0,0 +1,495 @@
+      partnerLink: PartnerLink,
+      user: User, // current user
+    ): Option[AppliedFragment] =
+      val up = LinkUser(targetProgram, targetUser, UserType.Standard, ProgramUserRole.CoiRO, partnerLink)
+      user.role match
+        case GuestRole                    => None
+        case ServiceRole(_)               => Some(up)
+        case StandardRole.Admin(_)        => Some(up)
+        case StandardRole.Ngo(_, partner) => Some(up |+| void" WHERE " |+| existsAllocationForPartner(targetProgram, partner))
+        case StandardRole.Staff(_)        => Some(up)
+        case StandardRole.Pi(_)           =>
+          Some(
+            up |+| void" WHERE " |+| existsUserAsPi(targetProgram, user.id)  |+|
+                   void" OR "    |+| existsUserAsCoi(targetProgram, user.id)
+          )
+
+    /**
+     * Link staff support to a program.
+     * - Staff, Admin, and Service users can always do this.
+     * - Nobody else can do this.
+     */
+    def linkSupport(
+      targetProgram: Program.Id,
+      targetUser: User.Id, // user to link
+      user: User, // current user
+      tpe: ProgramUserRole.SupportPrimary.type | ProgramUserRole.SupportSecondary.type
+    ): Option[AppliedFragment] =
+      import lucuma.core.model.Access._
+      val up = LinkUser(targetProgram, targetUser, UserType.Standard, tpe, PartnerLink.HasUnspecifiedPartner)
+      user.role.access match
+        case Admin | Staff | Service => Some(up) // ok
+        case _                       => None // nobody else can do this
+
+    def existsUserAs(
+      programId: Program.Id,
+      userId: User.Id,
+      role: ProgramUserRole
+    ): AppliedFragment =
+      sql"""
+        EXISTS (select c_role from t_program_user where c_program_id = $program_id and c_user_id = $user_id and c_role = $program_user_role)
+      """.apply(programId, userId, role)
+
+    def existsUserAsPi(
+      programId: Program.Id,
+      userId: User.Id,
+    ): AppliedFragment =
+      existsUserAs(programId, userId, ProgramUserRole.Pi)
+
+    def existsUserAsCoi(
+      programId: Program.Id,
+      userId: User.Id,
+    ): AppliedFragment =
+      existsUserAs(programId, userId, ProgramUserRole.Coi)
+
+    def existsAllocationForPartner(
+      programId: Program.Id,
+      partner: Partner
+    ): AppliedFragment =
+      sql"""
+        EXISTS (select c_duration from t_allocation where c_program_id = $program_id and c_ta_category=${lucuma.odb.util.Codecs.time_accounting_category} and c_duration > 'PT')
+      """.apply(programId, partner.timeAccountingCategory)
+
+    def existsUserAccess(
+      user:      User,
+      programId: Program.Id
+    ): Option[AppliedFragment] =
+      user.role match
+        case GuestRole                    => existsUserAsPi(programId, user.id).some
+        case StandardRole.Pi(_)           => (void"(" |+| existsUserAsPi(programId, user.id) |+| void" OR " |+| existsUserAsCoi(programId, user.id) |+| void")").some
```

Comment:
I think perhaps we may want to say that a CoI only has access if they have accepted an invitation?  At any rate, at the moment they will have access immediately if added via `addProgramUser`.